src/gun_pool.erl

```diff
@@ -0,0 +1,600 @@
+%% @todo Probably count degraded connections separately to avoid counting every time.
+degraded(info, Msg={gun_up, ConnPid, _}, StateData=#state{opts=Opts, conns=Conns}) ->
+	#{ConnPid := down} = Conns,
+	%% We optionally run the setup function if one is defined. The
+	%% setup function tells us whether we are fully up or not. The
+	%% setup function may be called repeatedly until the connection
+	%% is established.
+	%%
+	%% @todo It is possible that the connection never does get
+	%% fully established. We should deal with this. We probably
+	%% need to handle all messages.
+	{SetupFun, SetupState0} = setup_fun(Opts),
+	degraded_setup(ConnPid, Msg, StateData, SetupFun, SetupState0);
+%% @todo
+%degraded(info, Msg={gun_tunnel_up, ConnPid, _, _}, StateData0=#state{conns=Conns}) ->
+%	;
+degraded(info, Msg={gun_upgrade, ConnPid, _, _, _},
+		StateData=#state{opts=#{setup_fun := {SetupFun, _}}, conns=Conns}) ->
+	%% @todo Probably shouldn't crash if the state is incorrect, that's programmer error though.
+	#{ConnPid := {setup, SetupState0}} = Conns,
+	%% We run the setup function again using the state previously kept.
+	%% @todo We need to keep track of the StreamRef for Websocket I guess?
+	degraded_setup(ConnPid, Msg, StateData, SetupFun, SetupState0);
+degraded(Type, Event, StateData) ->
+	handle_common(Type, Event, ?FUNCTION_NAME, StateData).
+
+setup_fun(#{setup_fun := SetupFun}) ->
+	SetupFun;
+setup_fun(_) ->
+	{fun (_, {gun_up, _, Protocol}, _) ->
+		{up, Protocol, #{}}
+	end, undefined}.
+
+degraded_setup(ConnPid, Msg, StateData0=#state{conns=Conns, conns_meta=ConnsMeta}, SetupFun, SetupState0) ->
+	case SetupFun(ConnPid, Msg, SetupState0) of
+		Setup={setup, _SetupState} ->
+			StateData = StateData0#state{conns=Conns#{ConnPid => Setup}},
+			{keep_state, StateData};
+		%% The Meta is different from Settings. It allows passing around
+		%% Websocket or tunnel stream refs.
+		{up, Protocol, Meta} ->
+			Settings = #{},
+			StateData = StateData0#state{
+				conns=Conns#{ConnPid => {up, Protocol, Settings}},
+				conns_meta=ConnsMeta#{ConnPid => Meta}
+			},
+			case is_degraded(StateData) of
+				true -> {keep_state, StateData};
+				false -> {next_state, operational, StateData}
+			end
+	end.
+
+is_degraded(#state{conns=Conns0}) ->
+	Conns = maps:to_list(Conns0),
+	Len = length(Conns),
+	Ups = [up || {_, {up, _, _}} <- Conns],
+	Len =/= length(Ups).
+
+operational(Type, Event, StateData) ->
+	handle_common(Type, Event, ?FUNCTION_NAME, StateData).
+
+handle_common({call, From}, {get_connection, _ReqOpts}, _,
+		StateData=#state{conns_meta=ConnsMeta}) ->
+	case find_available_connection(StateData) of
+		none ->
+			{keep_state_and_data, {reply, From, undefined}};
+		ConnPid ->
+			Meta = maps:get(ConnPid, ConnsMeta, #{}),
+			{keep_state_and_data, {reply, From, {ConnPid, Meta}}}
+	end;
+handle_common(info, {gun_notify, ConnPid, settings_changed, Settings}, _, StateData=#state{conns=Conns}) ->
```

Comment:
We receive a message when HTTP/2 settings change. We keep track of the current max_concurrent_streams value this way.